### PR TITLE
executor: impl runaway watch check 

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: ref #43691

Problem Summary:

### What is changed and how it works?
ref #44339, we have introduced runawayChecker. In this PR, I add `BeforeExecutor` function to check  whether the query is in watch list.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
image

image

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/tests/resourcegroup/resource_group_test.go [hunk 1]
```diff
@@ -229,6 +229,11 @@ func TestResourceGroupRunaway(t *testing.T) {
 
 	tk.MustExec("set global tidb_enable_resource_control='on'")
 	tk.MustExec("create resource group rg1 RU_PER_SEC=1000 QUERY_LIMIT=(EXEC_ELAPSED='100ms' ACTION=KILL)")
+	tk.MustExec("create resource group rg2 BURSTABLE RU_PER_SEC=2000 QUERY_LIMIT=(EXEC_ELAPSED='50ms' action KILL WATCH EXACT duration '1m')")
+	tk.MustQuery("select * from information_schema.resource_groups where name = 'rg2'").Check(testkit.Rows("rg2 2000 MEDIUM YES EXEC_ELAPSED=50ms, ACTION=KILL, WATCH=EXACT[1m0s]"))
+	tk.MustQuery("select  /*+ resource_group(rg1) */ * from t").Check(testkit.Rows("1"))
```
Comment (Connor1996) [inline ddl/tests/resourcegroup/resource_group_test.go:234]: extra space after `select`

File: ddl/tests/resourcegroup/resource_group_test.go [hunk 2]
```diff
@@ -228,8 +228,13 @@ func TestResourceGroupRunaway(t *testing.T) {
...
 	tk.MustExec("set global tidb_enable_resource_control='on'")
-	tk.MustExec("create resource group rg1 RU_PER_SEC=1000 QUERY_LIMIT=(EXEC_ELAPSED='100ms' ACTION=KILL)")
-	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/store/copr/sleepCoprRequest", `sleep(500)`))
+	tk.MustExec("create resource group rg1 RU_PER_SEC=1000 QUERY_LIMIT=(EXEC_ELAPSED='50ms' ACTION=KILL)")
+	tk.MustExec("create resource group rg2 BURSTABLE RU_PER_SEC=2000 QUERY_LIMIT=(EXEC_ELAPSED='50ms' action KILL WATCH EXACT duration '1m')")
+	tk.MustQuery("select * from information_schema.resource_groups where name = 'rg2'").Check(testkit.Rows("rg2 2000 MEDIUM YES EXEC_ELAPSED=50ms, ACTION=KILL, WATCH=EXACT[1m0s]"))
+	tk.MustQuery("select  /*+ resource_group(rg1) */ * from t").Check(testkit.Rows("1"))
+	tk.MustQuery("select  /*+ resource_group(rg2) */ * from t").Check(testkit.Rows("1"))
+
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/store/copr/sleepCoprRequest", fmt.Sprintf("return(%d)", 60)))
```
Comment (CabinfeverB) [inline ddl/tests/resourcegroup/resource_group_test.go:237]: can accelerated test

File: errno/errname.go [hunk 3]
```diff
@@ -1134,11 +1134,12 @@ var MySQLErrName = map[uint16]*mysql.ErrMessage{
...
-	ErrResourceGroupConfigUnavailable: mysql.Message("Resource group configuration is unavailable", nil),
-	ErrResourceGroupThrottled:         mysql.Message("Exceeded resource group quota limitation", nil),
-	ErrResourceGroupQueryRunaway:      mysql.Message("Query execution was interrupted, identified as runaway query", nil),
+	ErrColumnInChange:                      mysql.Message("column %s id %d does not exist, this column may have been updated by other DDL ran in parallel", nil),
+	ErrResourceGroupSupportDisabled:        mysql.Message("Resource control feature is disabled. Run `SET GLOBAL tidb_enable_resource_control='on'` to enable the feature", nil),
+	ErrResourceGroupConfigUnavailable:      mysql.Message("Resource group configuration is unavailable", nil),
+	ErrResourceGroupThrottled:              mysql.Message("Exceeded resource group quota limitation", nil),
+	ErrResourceGroupQueryRunaway:           mysql.Message("Query execution was interrupted, identified as runaway query", nil),
+	ErrResourceGroupQueryRunawayQuarantine: mysql.Message("Quarantined and interrupted because of being watch list", nil),
```
Comment (Connor1996) [inline errno/errname.go:1142]: being in

Comment (Connor1996) [reply errno/errname.go:1142]: How about `Quarantined and interrupted because of being in runaway watch list`

**Review by Connor1996 — APPROVED** [review]
rest LGTM

File: ddl/tests/resourcegroup/resource_group_test.go [hunk 4]
```diff
@@ -241,6 +246,10 @@ func TestResourceGroupRunaway(t *testing.T) {
 
 	tk.MustExec("alter resource group rg1 RU_PER_SEC=1000 QUERY_LIMIT=(EXEC_ELAPSED='100ms' ACTION=DRYRUN)")
 	tk.MustQuery("select /*+ resource_group(rg1) */ * from t")
+
+	err = tk.QueryToErr("select /*+ resource_group(rg2) */ * from t")
+	require.ErrorContains(t, err, "Query execution was interrupted, identified as runaway query")
```
Comment (Connor1996) [inline ddl/tests/resourcegroup/resource_group_test.go:251]: make sure cooldown and dryrun are not killed

File: errno/errname.go [hunk 5]
```diff
@@ -1134,11 +1134,12 @@ var MySQLErrName = map[uint16]*mysql.ErrMessage{
...
-	ErrResourceGroupSupportDisabled:   mysql.Message("Resource control feature is disabled. Run `SET GLOBAL tidb_enable_resource_control='on'` to enable the feature", nil),
-	ErrResourceGroupConfigUnavailable: mysql.Message("Resource group configuration is unavailable", nil),
-	ErrResourceGroupThrottled:         mysql.Message("Exceeded resource group quota limitation", nil),
-	ErrResourceGroupQueryRunaway:      mysql.Message("Query execution was interrupted, identified as runaway query", nil),
+	ErrColumnInChange:                      mysql.Message("column %s id %d does not exist, this column may have been updated by other DDL ran in parallel", nil),
+	ErrResourceGroupSupportDisabled:        mysql.Message("Resource control feature is disabled. Run `SET GLOBAL tidb_enable_resource_control='on'` to enable the feature", nil),
+	ErrResourceGroupConfigUnavailable:      mysql.Message("Resource group configuration is unavailable", nil),
+	ErrResourceGroupThrottled:              mysql.Message("Exceeded resource group quota limitation", nil),
+	ErrResourceGroupQueryRunaway:           mysql.Message("Query execution was interrupted, identified as runaway query", nil),
```
Comment (nolouch) [inline errno/errname.go:1141]:
```suggestion
	ErrResourceGroupQueryRunawayInterrupted:           mysql.Message("Query execution was interrupted, identified as runaway query", nil),
```

**Review by nolouch — COMMENTED** [review]
rest lgtm

**Review by nolouch — APPROVED** [review]
lgtm

**Review by glorv — APPROVED** [review]
LGTM

**Pull request merged**